### PR TITLE
Migrate to jakarta namespace

### LINKED ISSUE #237: Spring boot 3 incompatibility
Hi! We are trying to upgrade our application to Spring Boot 3 and as part of the migration we had to move away from javax to jakarta as all the dependencies moved there. This creates incompatibility with this library since it is still based on javax implementations/classes. Question is, do you plan to release new version which migrates your implementation to jakarta from javax?

### PR DESCRIPTION
**Pull request by reda-alaoui**
Fixes #237
I don't know what to do with `httpunit`. Any advice? :)

### CODE REVIEW

**Comment by jetztgradnet** [comment]
We are currently migrating an application using this proxy servlet to JakartaEE and are looking for a solution as well. 
@reda-alaoui I will test your draft PR #238 and provide some feedback.
@dsmiley what would work for you regarding a timeline of a release, assuming the PR is ready and working?

File: pom.xml [hunk 1]
```diff
@@ -54,11 +54,10 @@
...
   <dependencies>
 
-    <!-- Probably works with 3.0 as well -->
     <dependency>
-      <groupId>javax.servlet</groupId>
-      <artifactId>javax.servlet-api</artifactId>
-      <version>3.1.0</version>
+      <groupId>jakarta.servlet</groupId>
+      <artifactId>jakarta.servlet-api</artifactId>
+      <version>5.0.0</version>
```
Comment (onderilkesever) [inline pom.xml:60]: I wonder why you used 5 instead of 6.0.0 which is latest?

Comment (reda-alaoui) [reply pom.xml:60]: For maximum compatibility. Some webapp containers only support `5.0.0`.

Comment (jetztgradnet) [reply pom.xml:60]: I agree that it should stay on 5.0 for JakartaEE 9 compatibility as e.g. used by Jetty 11. Any use in a JakartaEE 10 / Servlet 6 container should be backwards-compatible to that.

**Comment by stuplarosa** [comment]
We are also waiting for this change so that we can move to Tomcat 10.

**Comment by jetztgradnet** [comment]
@reda-alaoui regarding `httpunit`:

I found an updated version at [hazendaz/httpunit](https://github.com/hazendaz/httpunit) which works with JakartaEE.

So instead of 
```
    <dependency>
      <groupId>httpunit</groupId>
      <artifactId>httpunit</artifactId>
      <version>1.7</version>
      <scope>test</scope>
    </dependency>
```

use this:
```
    <dependency>
      <groupId>com.github.hazendaz.httpunit</groupId>
      <artifactId>httpunit</artifactId>
      <version>2.0.0</version>
      <scope>test</scope>
    </dependency>
```

**Comment by jetztgradnet** [comment]
I tested this locally and for me this worked. And the tests also run successfully with the updated httpunit variant mentioned in the previous comment.

**Comment by jetztgradnet** [comment]
@reda-alaoui may I suggest to update http-unit to the aforementioned version and publish this draft PR?

File: pom.xml [hunk 2]
```diff
@@ -54,11 +54,10 @@
 
   <dependencies>
 
-    <!-- Probably works with 3.0 as well -->
     <dependency>
-      <groupId>javax.servlet</groupId>
-      <artifactId>javax.servlet-api</artifactId>
-      <version>3.1.0</version>
+      <groupId>jakarta.servlet</groupId>
```
Comment (jetztgradnet) [inline pom.xml:58]: we also need to update/change `httpunit` to a variant supporting Jakarta EE:

```
<dependency>
      <groupId>com.github.hazendaz.httpunit</groupId>
      <artifactId>httpunit</artifactId>
      <version>2.0.0</version>
      <scope>test</scope>
    </dependency>
```

**Comment by EdgewareRoad** [comment]
Is the intent for this to be in a new release branch or fork of the code? Just aware that this will be a breaking change for anyone who can't update their Java or servlet engine?

Thanks

Phil

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by dsmiley** [comment]
I'll branch the current master branch to maybe "branch_1x" leaving the current master branch for the new 2x.

**Comment by EdgewareRoad** [comment]
Great. Many thanks David

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by jetztgradnet** [comment]
There are different ways to go about this. In paultuckey/urlrewritefilter#242 instead of migrating everything they use a Maven plugin to create a different `-jakarta` variant of the library by doing bytecode rewriting after the main build.

But IMHO in the long run, having a 2.0 release for the Jakarta-based servlet containers should be the right path.

**Comment by dsmiley** [comment]
That's a fantastic idea @jetztgradnet !  Basically use shading and a Maven classifier to ship it.  I didn't know it could be so easy.  This allows me to have it both ways -- support `javax.servlet` and also `jakarta.servlet` at the same time!  One branch of changes instead of porting to another branch or abandoning a branch and thus a subset of the user community.  Of course it won't be tenable forever; there may eventually be changes specific to the new Jakarta version but there has been none yet.

I'd like to go in that direction.  LMK if there are concerns.

**Comment by reda-alaoui** [comment]
@dsmiley as you can see at https://github.com/mitre/HTTP-Proxy-Servlet/pull/238/files#r1101714206 , this is not a simple package replacement.

**Comment by jetztgradnet** [comment]
@reda-alaoui right, but adding the shading plus changing this line is still possible, right? the `setStatus(int)` method has been there al the time, it's only the variant with the additional message line that was remove.

**Comment by jetztgradnet** [comment]
@dsmiley whether with a new major version for jakrata-only or a dual-variant approach, it would be great to get a release version soon :-)

**Comment by jetztgradnet** [comment]
So we now have two possible approaches:

1. provide two variants of the library, the main one unchanged and a variant with classifier `jakarta`, as a `1.13` release
2. leave the `1.x` branch unchanged and migrate the main branch to Jakarta EE with 2.0 release providing only this way, i.e. essentially this PR provided by @reda-alaoui

@dsmiley so which way do you want to go? I am following @reda-alaoui argument, but at the same time it is my main short-term interest in getting a release supporting Jakarta EE with any of the two approaches:

If you prefer the first approach, I would be willing to provide a PR with that noe-line change required to work for Servlets 6.x plus the shading (assuming we can simply transfer the approach from paultuckey/urlrewritefilter#242).

**Comment by dsmiley** [comment]
Then we do a bit of both.  Embrace the "shade" mechanism to allow one maintained codeline and release but also do everything this PR is doing (new major version; use Jakarta in the maintained code).  The "shade" configuration would be to produce a "javax-servlet" classifier (or something like that; dunno if there's a standard).

**Comment by jetztgradnet** [comment]
That sounds like a great compromise. In the simplest case we can take the example from the urlrewritefilter and reverse 'javax.servlet' and 'jakarta.servlet'. 

@reda-alaoui would you also do that or would you want me to provide a snippet for that?

@dsmiley any preference for a classifier name? 'javax' or 'javax.servlet'

**Comment by dsmiley** [comment]
I guess `javax` would be more generalized as the overall matter isn't just servlets.  It's shorter too.

**Comment by jetztgradnet** [comment]
So I didn't want to re-create @reda-alaoui's PR or provide a different PR include that commit, so adding this snippet to `pom.xml` after line 155 produces both jars when building the project using `maven source:jar install`:

```
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-shade-plugin</artifactId>
        <version>3.2.2</version>
        <executions>
          <execution>
            <id>javax</id>
            <phase>package</phase>
            <goals>
              <goal>shade</goal>
            </goals>
            <configuration>
              <shadedClassifierName>javax</shadedClassifierName>
              <shadedArtifactAttached>true</shadedArtifactAttached>
              <createDependencyReducedPom>false</createDependencyReducedPom>
              <artifactSet>
                  <includes>
                    <include>${project.groupId}:${project.artifactId}</include>
                  </includes>
              </artifactSet>
              <relocations>
                <relocation>
                  <pattern>jakarta.servlet</pattern>
                  <shadedPattern>javax.servlet</shadedPattern>
                </relocation>
                <relocation>
                  <pattern>jakarta.annotation</pattern>
                  <shadedPattern>javax.annotation</shadedPattern>
                </relocation>
              </relocations>
            </configuration>
          </execution>
        </executions>
      </plugin>
```

So @reda-alaoui or @dsmiley, whoever is first, would either of you add this commit to this PR or afterwards after merging this PR? 
And hopefully @dsmiley would then release a binarty artifact on Maven Central (after possibly adding a paragraph to the README pointing out support for both versions)? ;-)

**Comment by reda-alaoui** [comment]
@jetztgradnet @dsmiley I added the shading plugin

**Comment by jetztgradnet** [comment]
Hey @dsmiley,
did you have a chance looking at creating a release yet? Anything I could do to help?
Thanks for your time, much appreciated!

**Comment by dsmiley** [comment]
Well I see test failures still.  Can they be addressed please?

**Comment by sconrads** [comment]
Any progress on this PR? If I could help in any way, please let me know :)

**Comment by plan3d** [comment]
The problem seems to occur only with httpclient versions 4.3.x and 4.4.x. Is it still necessary to support/test these versions? Build with httpclient version 4.5.14 works on my machine. Can versions 4.3.6 and 4.4.1 be removed from .github/workflows/maven.yml?

**Comment by dsmiley** [comment]
Sure; we can drop support for older HttpClient releases if this is what it takes.

**Comment by jetztgradnet** [comment]
@plan3d are you going to remove those old httpclient version or would you prefer if somebody else jumps in? I didn't want to just push another commit to this PR without checking with you first...

**Comment by jetztgradnet** [comment]
I assume removig support/testing for old httpclient versions would mean to simply remove those lines from `.github/workflows/maven.yml`?

**Comment by jetztgradnet** [comment]
@reda-alaoui any comment from your side on removing the old httpclient tests?
Would be great to get this finally closed and released... :-)

**Comment by reda-alaoui** [comment]
On it.

**Comment by reda-alaoui** [comment]
Done. Looks like someone has to trigger the build.

**Review by dsmiley — COMMENTED** [review]
Thanks everyone!

The CHANGES.md needs to be updated as does README.md.  Can that be done please?  README references older HttpClient support which is now false.  Needn't be thorough but don't want wrong information.  README should reference the dual classifier thing.

**Comment by reda-alaoui** [comment]
@dsmiley CHANGES.md and README.md updated.

**Comment by dsmiley** [comment]
Please don't force-push changes to PRs.  It means reviewers (like me) can't see what changes you did between now and before.

**Comment by reda-alaoui** [comment]
> Please don't force-push changes to PRs. It means reviewers (like me) can't see what changes you did between now and before.

Ok noted. Some repository want 1 commit per PR. Others don't. That's why I prefer [Gerrit](https://www.gerritcodereview.com/) over Github.

**Comment by dsmiley** [comment]
Interesting... never seen such a 1-commit per PR policy.  In GitHub, it's trivial to view all changes collapsed -- it's the default view if you click the "Files Changed" view.  So I don't get the point.

File: README.md [hunk 3]
```diff
@@ -27,8 +27,20 @@ In the meantime, you'll have to add the jar files for this and its dependencies:
...
         |  \- commons-codec:commons-codec:jar:1.11:compile
 
-This proxy supports HttpClient 4.3, and newer version too.
-If you need to support _older_ HttpClient versions, namely 4.1 and 4.2, then use  the 1.8 version of this proxy.
+This proxy supports HttpClient 4.5, and newer version too.
+If you need to support _older_ HttpClient versions:
+* use 1.8 version of this proxy for HttpClient versions 4.1 and 4.2
+* use 1.12 version of this proxy for HttpClient versions 4.3 and 4.4
+
+As of version 2.0 of the proxy, `jakarta servlet-api` is supported by default. If you still need to support `javax servlet-api`, you must specify the `javax` classifier on the dependency declaration as follows:
```
Comment (dsmiley) [inline README.md:35]:
```suggestion
As of version 2.0 of the proxy, the proxy switched to the `jakarta servlet-api`, while nonetheless retaining support for the `javax servlet-api` version for those that need it.  To use that, specify the `javax` classifier on the dependency declaration as follows:
```

**Comment by dsmiley** [comment]
I noticed Jakarta Servlet spec 5.0 was introduced here but that there are newer specs like 6 (judging from a question in #242 ) that was raised today.  Is there an opinion here for/against raising the spec level, and to what?

**Comment by reda-alaoui** [comment]
@dsmiley , I think [my comment](https://github.com/mitre/HTTP-Proxy-Servlet/pull/238/files#r1085119614) still stands.

**Pull request merged**